lib/services/profiles-api-service.js

```diff
@@ -102,41 +102,80 @@ function profileApiServiceFactory(
 
                 });
             } else {
-                return self.createNodeAndRunDiscovery(macAddresses);
+                return self.createNodeAndRunDiscovery(macAddresses, options);
             }
         });
     };
 
-    ProfileApiService.prototype.runDiscovery = function(node) {
+    ProfileApiService.prototype.runDiscovery = function(node, options) {
         var self = this;
+        var configuration;
 
+        if (node.type === 'switch') {
+            configuration = self.getSwitchDiscoveryConfiguration(node, options.switchVendor);
+        } else {
+            configuration = {
+                name: 'Graph.SKU.Discovery',
+                options: {
+                    defaults: {
+                        graphOptions: {
+                            target: node.id
+                        },
+                        nodeId: node.id
+                    }
+                }
+            };
+        }
+
+        // The nested workflow holds the lock against the nodeId in this case,
+        // so don't add it as a target to the outer workflow context
+        return workflowApiService.createAndRunGraph(configuration, null)
+        .then(function() {
+            return self.waitForDiscoveryStart(node.id);
+        })
+        .then(function() {
+            return node;
+        });
+    };
+
+    ProfileApiService.prototype.getSwitchDiscoveryConfiguration = function(node, vendor) {
         var configuration = {
-            name: 'Graph.SKU.Discovery',
+            name: 'Graph.SKU.Switch.Discovery.Active',
             options: {
                 defaults: {
                     graphOptions: {
                         target: node.id
                     },
                     nodeId: node.id
+                },
+                'vendor-discovery-graph': {
+                    graphName: null
                 }
             }
         };
 
-        return workflowApiService.createAndRunGraph(configuration)
-        .then(function() {
-            return self.waitForDiscoveryStart(node.id);
-        })
-        .then(function() {
-            return node;
-        });
+        vendor = vendor.toLowerCase();
+
+        if (vendor === 'cisco') {
```

Comment:
Could there be a generic script that is executed on the switch and pulls down it's relevant vendor specific script? (via downloadUrl maybe?).  That way all the vendor checks are contained to the switch and we still avoid munging scripts?